### ISSUE TITLE
[master] KWM huge memory usage

### ISSUE THREAD

**Issue by osatel** [koekeishiya/kwm#540]
After running KWM for some hours I experience it using a huge amount of memory. 

http://i.imgur.com/9MhMp8u.png

I tried installing it with brew first, exact same thing happened.
not quite sure why this occurs. 

or does the program require that much memory?

EDIT: i'm on MacOS Sierra 10.12.1

EDIT2: http://i.imgur.com/DrLh5TL.png

it doesn't add up though, so i'm not sure why activity monitor is registering so much ram out of kwm...

**Comment by chrishoage** [comment]
![screen shot 2016-12-02 at 10 22 28 am](https://cloud.githubusercontent.com/assets/1056886/20845300/41a430de-b879-11e6-9a51-a59257af69a0.png)

For reference there is mine running under El Cap. Maybe there is an issue with a memory leak in Sierra? 

(this has been running for _week_ on my machine)

**Comment by dominiklohmann** [comment]
<img width="541" alt="screen shot 2016-12-02 at 19 29 04" src="https://cloud.githubusercontent.com/assets/4488655/20845490/a2c9c6ac-b8c5-11e6-8d24-3dab267abd74.png">

Uptime is slightly over a day. Also on Sierra 10.12.1 (the build with Touch Bar support).

What does your config look like? Maybe we can reproduce it with your `kwmrc`.

**Comment by choco** [comment]
I actually have similar memory usage as @osatel  with 2 days uptime, that's super strange :/

**Comment by osatel** [comment]
@dominiklohmann I was using the example config.

**Comment by Jarvvski** [comment]
I too am getting a massive kwm load

![](https://i.imgur.com/r3Yuprk.png)

```
// Set default values for screen padding
kwmc config padding 40 20 20 20

// Set default values for container gaps
kwmc config gap 15 15

// Default tiling mode for Kwm (bsp | monocle | float)
kwmc config tiling bsp

// Let Kwm listen for hotkeys
kwmc config hotkeys on

// Automatically float windows that fail to resize
kwmc config float-non-resizable on

// Automatically resize the window to its container
kwmc config lock-to-container on

// Set focus-follows-mouse-mode to autoraise
kwmc config focus-follows-mouse on

/* Focus-follows-mouse is temporarily disabled when
   a floating window has focus */
kwmc config standby-on-float on

/* The mouse will automatically move to the center
   of the focused window */
kwmc config mouse-follows-focus on
// Allow window focus to wrap-around
kwmc config cycle-focus on

/* Override default settings for space 1 on screen 0
   kwmc config space 0 1 mode monocle
   kwmc config space 0 1 padding 100 100 100 100
   kwmc config space 0 1 gap 40 40 */

/* Override default settings for screen 1
   kwmc config display 1 mode bsp
   kwmc config display 1 padding 100 100 100 100
   kwmc config display 1 gap 40 40 */

// Set default container split-ratio
kwmc config split-ratio 0.5

// New splits become the left leaf-node
kwmc config spawn left

/* Add custom tiling rules for applications that
   does not get tiled by Kwm by default.
   This is because some applications do not have the
   AXWindowRole and AXStandardWindowSubRole */
kwmc rule owner="iTerm2" properties={role="AXDialog"}

// Enable border for focused window
kwmc config border focused on
kwmc config border focused size 1
kwmc config border focused color 0xFFFFFF
kwmc config border focused radius 0

// Enable border for marked window
kwmc config border marked on
kwmc config border marked size 1
kwmc config border marked color 0xFF336699
kwmc config border marked radius 0

// Enable the prefix mode
kwmc mode prefix prefix on
kwmc mode prefix timeout 0.75
kwmc mode prefix restore default
kwmc mode prefix color 0xFF458588

/*  Default Keybinds */
// Activate prefix mode
kwmc bindsym ctrl-s mode activate prefix

// Quit Kwm
kwmc bindsym cmd+alt+ctrl-q quit

// Launch iTerm2
kwmc bindsym cmd-return exec open -na /Applications/iTerm.app

// Rotate Window-Tree By 90degrees (Clockwise)
kwmc bindsym cmd+ctrl-r tree rotate 90

// Modify Container
kwmc bindsym prefix-h window -c reduce 0.05
kwmc bindsym prefix-l window -c expand 0.05

// Set Temporary Window Container
kwmc bindsym prefix-f window -z fullscreen
kwmc bindsym prefix-d window -z parent
kwmc bindsym prefix-w window -t focused

// Mark Window
kwmc bindsym cmd+alt+ctrl-m window -mk focused
kwmc bindsym cmd+alt+ctrl-h window -mk west wrap
kwmc bindsym cmd+alt+ctrl-l window -mk east wrap
kwmc bindsym cmd+alt+ctrl-j window -mk south wrap
kwmc bindsym cmd+alt+ctrl-k window -mk north wrap

// Give Focus To Window
kwmc bindsym cmd+alt-h window -f west
kwmc bindsym cmd+alt-l window -f east
kwmc bindsym cmd+alt-j window -f south
kwmc bindsym cmd+alt-k window -f north

kwmc bindsym cmd+alt-n window -fm prev
kwmc bindsym cmd+alt-m window -fm next

// Give Focus To Screen
kwmc bindsym cmd+alt-1 display -f 0
kwmc bindsym cmd+alt-2 display -f 1
kwmc bindsym cmd+alt-3 display -f 2

// Swap Focused Window
kwmc bindsym ctrl+alt-h window -s west
kwmc bindsym ctrl+alt-j window -s south
kwmc bindsym ctrl+alt-k window -s north
kwmc bindsym ctrl+alt-l window -s east
kwmc bindsym ctrl+alt-m window -s mark

// Pseudo containers
kwmc bindsym cmd+ctrl+alt-p tree -pseudo create
kwmc bindsym cmd+ctrl+alt-o tree -pseudo destroy
kwmc bindsym ctrl+alt-p window -s prev
kwmc bindsym ctrl+alt-n window -s next

// Detach Focused Window And Reinsert In Direction
kwmc bindsym ctrl+shift-k window -m north
kwmc bindsym ctrl+shift-l window -m east
kwmc bindsym ctrl+shift-j window -m south
kwmc bindsym ctrl+shift-h window -m west

// Detach Marked Window And Reinsert At Focused Window
kwmc bindsym ctrl+shift-x window -m mark

// Move Focused Window To Space
kwmc bindsym ctrl+alt-left window -m space left
kwmc bindsym ctrl+alt-right window -m space right

// Move Focused Window To Screen
kwmc bindsym ctrl+alt-1 window -m display 0
kwmc bindsym ctrl+alt-2 window -m display 1
kwmc bindsym ctrl+alt-3 window -m display 2

// Increase Container Gaps
kwmc bindsym prefix-x space -g increase horizontal
kwmc bindsym prefix-y space -g increase vertical

// Decrease Container Gaps
kwmc bindsym prefix+shift-x space -g decrease horizontal
kwmc bindsym prefix+shift-y space -g decrease vertical

// Increase Screen Padding
kwmc bindsym prefix-left space -p increase left
kwmc bindsym prefix-right space -p increase right
kwmc bindsym prefix-up space -p increase top
kwmc bindsym prefix-down space -p increase bottom
kwmc bindsym prefix-p space -p increase all

// Decrease Screen Padding
kwmc bindsym prefix+shift-left space -p decrease left
kwmc bindsym prefix+shift-right space -p decrease right
kwmc bindsym prefix+shift-up space -p decrease top
kwmc bindsym prefix+shift-down space -p decrease bottom
kwmc bindsym prefix+shift-p space -p decrease all
```

**Comment by osatel** [comment]
I've noticed it seems very stable before 1 day of uptime, it stays between 70 and 200 mb.
After about 1 day of uptime it starts ramping up hard.

**Comment by koekeishiya** [comment]
There is for sure a memory leak happening when a display is disconnected.
I am not aware of any other memory leaks, but Kwm should not at all, ever, require more than 100mb (theoretically at least). It's usually sitting at ~25mb for me.

I have been profiling memory usage for some time, and there are no reported leaks (in whatever parts I use). I do suspect there are some use cases where memory has been allocated that we no longer use, but still keep track of, for whatever reason.

**Comment by f1337** [comment]
Same issue here. I use both my MBP display and an external display, and it has been disconnected and reconnected at least once.

![screen shot 2016-12-21 at 1 29 36 pm](https://cloud.githubusercontent.com/assets/92846/21406755/11a52934-c782-11e6-8ecb-ec24a118906e.png)

**Comment by koekeishiya** [comment]
So after disabling the code for borders, memory usage went down to 4-5mb.
Will take a closer look when I have time.

**Comment by jongold** [comment]
@koekeishiya thanks - disabling borders fixed it for me too :)

**Comment by aeikenberry** [comment]
My kwm memory usage was over 3GB as well. I noticed, that if i open and close a window, I can see the RAM usage climbing each time. I have a space with 3 windows, if i open activity manager, it puts it nicely in the grid, i can note the ram, close AM, then press cmd+1 to open that window again, and KWM will visibly jump up, climbing every time I perform that.

**Comment by koekeishiya** [comment]
Memory usage of borders should be drastically lower after the latest commit.
Please do post back after some testing.

Thanks.

**Comment by dominiklohmann** [comment]
I've tested this for a few days now and have not encountered any memory issues since with borders running. This should probably prompt a 4.0.4 bug fix release.

**Comment by koekeishiya** [comment]
Thanks for the response, will do a release shortly.

**Comment by koekeishiya** [comment]
I am going to consider this issue fixed in version 4.0.4.
Please do not hesitate to create a new issue if the problem persists.
